### PR TITLE
[IMP] mail: Chatter Search + Emojis

### PR DESCRIPTION
**Pull request by SimonGenin**
Add search feature to messages
Refactor the code that handle the content of messages before render.

Task-2359037

Add Emojis and emojis popover.

### CODE REVIEW

File: addons/mail/static/src/models/message/message.js [hunk 1]
```diff
@@ -445,9 +510,158 @@ function factory(dependencies) {
...
+         * @param {String} hash 
+         */
+        readLessElementClicked(hash) {
+            for (let i = 0; i < this.currentlyOpenedReadMoreHashes.length; i++) { 
+                if (this.currentlyOpenedReadMoreHashes[i] === hash) { 
+                    this.currentlyOpenedReadMoreHashes.splice(i, 1); 
+                    i--; 
+                }
+            }
+        }
```
Comment (sdegueldre) [inline addons/mail/static/src/models/message/message.js:624]: This really makes it looks like you should be using a Set instead of an array.

Comment (seb-odoo) [reply addons/mail/static/src/models/message/message.js]: Or even better a new model and an x2many relation instead of an array? Relations are automatically Set and properly trigger render on update without doing array tricks.

Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js]: @seb-odoo Isn't a model for just a set of values a bit too much. That would create a new file that with a somewhat obscur name. Doesn't seem like it's a good idea in this case imo.

File: addons/mail/static/src/components/message/message.js [hunk 2]
```diff
@@ -414,22 +277,23 @@ export class Message extends Component {
...
+        }
+        if (this._subtypeDescription.el) {
+            this._subtypeDescription.el.innerHTML = this.message.processDomTransformation(this.message.subtype_description);
         }
+        if (this._trackingValuesRef.el) {
+            const trackingValuesDomElements = this._trackingValuesRef.el.querySelectorAll('[data-tracking-value]');
+            for (const element of trackingValuesDomElements) {
+                element.innerHTML = this.message.processDomTransformation(element.dataset.trackingValue);
+            }
+        }
```
Comment (SimonGenin) [inline addons/mail/static/src/components/message/message.js:295]: Maybe use refs ? See how it works

Comment (SimonGenin) [inline addons/mail/static/src/components/message/message.js:295]: All these could be fields to handle the caching. However, need to find a trick for the tracking values. Or just an field with default: {}.

File: addons/mail/static/src/components/message/message.js [hunk 3]
```diff
@@ -414,22 +269,26 @@ export class Message extends Component {
         if (!this.message) {
             return;
         }
-        if (this._prettyBodyRef.el && this.message.prettyBody !== this._lastPrettyBody) {
+        
+        if (this._authorRef.el) {
+            this._authorRef.el.innerHTML = this.message.processDomTransformation(this.message.author.nameOrDisplayName);
```
Comment (mart-e) [inline addons/mail/static/src/components/message/message.js:274]: xss?

File: addons/mail/static/src/components/message/message.js (hunk 2 above)
Comment (SimonGenin) [reply addons/mail/static/src/components/message/message.js]: done

File: addons/mail/static/src/components/message/message.js (hunk 3 above)
Comment (SimonGenin) [reply addons/mail/static/src/components/message/message.js]: I'll double check, but all the data should originate from the server.

Comment (mart-e) [reply addons/mail/static/src/components/message/message.js]: So, you still can't trust it. Rename your user `` and send a message to the admin user. Same for the subject. See the warning at https://www.odoo.com/documentation/14.0/developer/reference/addons/security.html#escaping-vs-sanitizing

File: addons/mail/static/src/models/message/message.js [hunk 4]
```diff
@@ -245,6 +267,30 @@ function factory(dependencies) {
...
+         * It uses a string representation of the DOM as we don't manipulate directly the real DOM with OWL. 
+         * 
+         * @param {String} content dom string rep on which to apply all the dom maniplation
+         * @returns processed content
+         */
+        processDomTransformation(content, isSafe = false) {
+            if (!content) return "";
+            if (!isSafe) {
+                content = _.escape(content);
+            }
```
Comment (SimonGenin) [inline addons/mail/static/src/models/message/message.js:281]: @mart-e All the data injected through the .innerHTML attribute go through this function. 
Would a simple escaping here like I did be enough ? It fixes the example your provided, but I'm not familiar with other ways it could get used.

File: addons/mail/static/src/components/thread_search_box/thread_search_box.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     * Updates the content
+     *
+     * @private
+     */
+    _update() {
+        if (!this.thread) {
+            return;
+        }
+        this._inputRef.el.value = this.thread.searchedText;
+    }
```
Comment (sdegueldre) [inline addons/mail/static/src/components/thread_search_box/thread_search_box.js:62]: lot of plumbing, can't you use t-model for this?

File: addons/mail/static/src/components/message/message.xml [hunk 6]
```diff
@@ -115,23 +113,23 @@
                             hasReplyIcon="props.hasReplyIcon"
                             hasMarkAsReadIcon="props.hasMarkAsReadIcon"
                         />
-                        <div class="o_Message_content m-2" t-ref="content">
-                            <div class="o_Message_prettyBody" t-ref="prettyBody"/><!-- message.prettyBody is inserted here from _update() -->
```
Comment (sdegueldre) [inline addons/mail/static/src/components/message/message.xml:119]: probably want to keep that comment somewhere (and update its content)

File: addons/mail/static/src/components/message/message.js [hunk 7]
```diff
@@ -39,15 +41,27 @@ export class Message extends Component {
          * to decide if it has to be updated.
          */
         this._lastPrettyBody;
```
Comment (sdegueldre) [inline addons/mail/static/src/components/message/message.js:43]: Looks like this is now unused and has been replaced by contentCache?

File: addons/mail/static/src/models/message/message.js [hunk 8]
```diff
@@ -523,6 +648,156 @@ function factory(dependencies) {
...
+         * @param {String} content
+         * @returns {String}
+         */
+        _highlightBasedOnThreadSearch(content) {
+            if (this.threads) {
+                for (const thread of this.threads) {
+                    if (thread.searchedText) {
+                        content = content.replace(
+                            new RegExp(`(${thread.searchedText})`,"ig"),
+                            '<b class="o_MessageInlineHighlight">$1</b>'
```
Comment (sdegueldre) [inline addons/mail/static/src/models/message/message.js:664]: You can absolutely not do text-search over the html as text and expect it to work. Tag names, attributes and a whole lot of other stuff *will* match and break.

File: addons/mail/static/src/models/message/message.js (hunk 4 above)
Comment (mart-e) [reply addons/mail/static/src/models/message/message.js:281]: Yes, it fixes the problem here. As a general rule, you should avoid using `innerHtml` (use `textContent`) or constructing HTML by hand (use `createElement`, `setAttribute`, etc.) whenever possible.  Always be defensive and assume the content can not be trusted. Don't forget to include HTML in your test cases to make sure the escaping is correctly done once (and zero or twice).

File: addons/mail/static/src/models/message/message.js [hunk 9]
```diff
@@ -523,6 +648,156 @@ function factory(dependencies) {
...
+         * Wraps emojis with a span to give them a class for formatting. 
+         * 
+         * @private
+         * @param {String} content
+         * @returns {String}
+         */
+        _wrapEmojisForFormatting(content) {
+            let processedContent;
+            for (const emoji of emojis) {
+                const { unicode } = emoji;
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js:780]: This code has not be changed, and is to be on the emoji task so I'll leave it for now.

File: addons/mail/static/src/models/thread_cache/thread_cache.js [hunk 10]
```diff
@@ -89,6 +89,93 @@ function factory(dependencies) {
...
+         * @private
+         * @returns {mail.message[]}
+         */
+        _computeFilteredMessages() {
+            if (!this.thread) {
+                return [['unlink-all']];
+            }
+            let searchedText = this.thread.searchedText;
+            if (!searchedText || /<.+>/g.exec( searchedText )) {
+                return [['unlink-all']];
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/mail/static/src/models/thread_cache/thread_cache.js:108]: Ho I didn't know it existed. The previous owner of the PR wrote this bit before it was introduced I believe.

File: addons/mail/static/src/models/message/message.js (hunk 8 above)
Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js:664]: @sdegueldre I agree. It was decided for a v1 that we didn't care, but that's just bad. Then the best solution I see is:
1) construction a DOMElement
2) Get all the text node and text content
3) Apply the transformation on those
4) Return the innerHTML from the constructed Element

Slower but only used when searching so I guess it's kinda alright 
Thoughts ?

File: addons/mail/static/src/models/message/message.js [hunk 11]
```diff
@@ -523,6 +648,156 @@ function factory(dependencies) {
...
+         * When opening a collapsable content, should it open all the children as well ?
+         * Making the effective depth of the collapsable equals 1.
+         * @returns String rep of the DOM Content processed
+         */
+        _insertMoreOrLessTogglersForCollapsableContent(content, shouldOpenChidlrenRecursively = false) {
+            /**
+             * For simplicity, we need to manipulate a DOM Element directly.
+             * Note that "unvalid" html, like nested <p> will get unested by this procedure.
+             * So the better the html conventions are respected, the better this will work.
+             */
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js:686]: @sdegueldre I got the same result using the DOMParser. It tries to unnest it as well (and the result is not pretty...)

File: addons/mail/static/src/models/message/message.js [hunk 12]
```diff
@@ -523,6 +648,156 @@ function factory(dependencies) {
...
+        /**
+         * Adds Read More and Read Less togglers for the collapsable contents.
+         * 
+         * @param {String} content String rep of the DOM Content
+         * @param {Boolean} shouldOpenChidlrenRecursively 
+         * When opening a collapsable content, should it open all the children as well ?
+         * Making the effective depth of the collapsable equals 1.
+         * @returns String rep of the DOM Content processed
+         */
+        _insertMoreOrLessTogglersForCollapsableContent(content, shouldOpenChidlrenRecursively = false) {
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js:681]: @sdegueldre To be discussed with @seb-odoo and @alexkuhn I guess. But when developing this, I did have indeed that feeling the only use case I came across was rather simple but the code tries to manage much more complex possibilities. I think we could rethink what we want to do with that on a functional perspective.

File: addons/mail/static/src/models/message/message.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/mail/static/src/models/message/message.js:693]: I'm not getting that on either chrome or firefox, are you using the correct mimetype? (`text/xml`)

Although since we're going to end up inserting this in the page it doesn't matter all that much, it will 'unnest' at that point, I just thought I'd mention it in passing.

File: addons/mail/static/src/models/message/message.js (hunk 8 above)
Comment (sdegueldre) [reply addons/mail/static/src/models/message/message.js:697]: Sorry for the late reply, I think I missed the notification.

Yes, I think that's better, using a naive approach will probably have some limitations (eg no matches if the text is partially bold/italics) but at least it won't just break the existing markup. There are ways to improve this further later (eg by creating a string containing the full text content, and then finding match offsets in the string and mapping that back to DOM elements) but I think as a first approach it should be enough.

Still not a fan of going through html-as-text every time but I guess currently there's not really a good way around it.

Comment (SimonGenin) [reply addons/mail/static/src/models/message/message.js:703]: I went with a smarter regex cause the HTML manipulation solution came out to be more tedious than I though. I tried a few methods and each time the result was not good and the code quite complex.

File: addons/mail/static/src/models/thread_cache/thread_cache.js [hunk 13]
```diff
@@ -89,6 +89,87 @@ function factory(dependencies) {
...
+                return false;
+            }
+            // If the data has changed, it means we're done
+            if (this._filteredMessages !== this.filteredMessages) {
+                this._filteredMessages = this.filteredMessages;
+                return false;
+            } 
+            // Otherwise, if the data has not changed, we're still pending. 
+            return true;
+        }
```
Comment (SimonGenin) [inline addons/mail/static/src/models/thread_cache/thread_cache.js:171]: I have strong doubts this works. In reality, while the search is client side, having a loader is useless. But when it will be server side, then it will be important. And there will be a code flux to actually easily set IsMessageSearchPending without a compute.
